trunner/device.py

```diff
@@ -30,13 +33,228 @@
-            break
-        if output:
-            logging.info(output)
+def is_github_actions():
+    return os.getenv('GITHUB_ACTIONS', False)
+
+
+class Psu:
+    """Wrapper for psu program"""
+
+    def __init__(self, script, cwd=_BOOT_DIR):
+        self.script = script
+        self.cwd = cwd
+        self.proc = None
+
+    def read_output(self):
+        if is_github_actions():
+            logging.info('::group::psu\n')
+
+        while True:
+            line = self.proc.stdout.readline().decode('utf-8')
+            if not line:
+                break
+
+            logging.info(line)
+
+        if is_github_actions():
+            logging.info('::endgroup::\n')
+
+    def run(self):
+        self.proc = subprocess.Popen(
+            ['psu', f'{self.script}'],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.DEVNULL,
+            cwd=self.cwd
+        )
+
+        self.read_output()
+        self.proc.wait()
+        if self.proc.returncode != 0:
+            logging.error(f'Command {" ".join(self.proc.args)} with pid {self.proc.pid} failed!\n')
+            raise Exception('Flashing IMXRT106x failed')
+
+
+class PhoenixdError(Exception):
+    pass
+
+
+class Phoenixd:
+    """ Wrapper for phoenixd program"""
+
+    def __init__(
+        self,
+        port,
+        baudrate=460800,
+        dir='.',
+        cwd=_BOOT_DIR,
+        wait_dispatcher=True
+    ):
+        self.port = port
+        self.baudrate = baudrate
+        self.dir = dir
+        self.cwd = cwd
+        self.proc = None
+        self.reader_thread = None
+        self.wait_dispatcher = wait_dispatcher
+        self.dispatcher_event = None
+        self.output_buffer = ''
+
+    def _reader(self):
+        """ This method is intended to be run as a separated thread. It reads output of proc
+            line by line and saves it in the output_buffer. Additionally, if wait_dispatcher is true,
+            it searches for a line stating that message dispatcher has started """
+
+        while True:
+            line = self.proc.readline()
+            if not line:
+                break
+
+            if not self.dispatcher_event.is_set():
+                msg = f'Starting message dispatcher on [{self.port}] (speed={self.baudrate})'
+                if msg in line:
+                    self.dispatcher_event.set()
+
+            self.output_buffer += line
+
+    def run(self):
+        # Use pexpect.spawn to run a process as PTY, so it will flush on a new line
+        self.proc = pexpect.spawn(
+            'phoenixd',
+            ['-p', self.port,
+             '-b', str(self.baudrate),
+             '-s', self.dir],
+            cwd=self.cwd,
+            encoding='utf-8'
+        )
+
+        self.dispatcher_event = threading.Event()
+        if not self.wait_dispatcher:
+            # If we don't want to wait for a message dispatcher, just mark it as started
+            self.dispatcher_event.set()
```

Comment:
[nit] would be more readable if `if self.wait_dispather` block would be around lines 138-143

Reply:
I agree, done!